### ISSUE TITLE
When starting a build, abstruse die

### ISSUE THREAD

**Issue by marcopiraccini** [bleenco/abstruse#377]
Docker: `version 17.12.1-ce, build 7390fc6`
Ubuntu: Ubuntu 18.04 x64
Node: v6.14.2

When starting a build, the application die: 

```
[2018-05-13 11:59:25]: [build]: build: 20 job: 20 => job started
[2018-05-13 11:59:26]: [http]: request to xxxxx successful
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: write after end
    at writeAfterEnd (_stream_writable.js:195:12)
    at Writable.write (_stream_writable.js:242:5)
    at IncomingMessage.<anonymous> (/root/.nvm/versions/node/v6.14.2/lib/node_modules/abstruse/node_modules/docker-modem/lib/modem.js:301:16)
    at emitNone (events.js:86:13)
    at IncomingMessage.emit (events.js:185:7)
    at emitReadable_ (_stream_readable.js:432:10)
    at emitReadable (_stream_readable.js:426:7)
    at readableAddChunk (_stream_readable.js:187:13)
    at IncomingMessage.Readable.push (_stream_readable.js:134:10)
    at HTTPParser.parserOnBody (_http_common.js:123:22)
    at Socket.socketOnData (_http_client.js:367:20)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at Pipe.onread (net.js:559:20)
```

**Comment by marcopiraccini** [comment]
Same with node `v8.11.1`

**Comment by marcopiraccini** [comment]
Here were try to write on a closed stream: https://github.com/apocas/docker-modem/blob/master/lib/modem.js#L301

**Comment by marcopiraccini** [comment]
Just for troubleshooting:
- When this happens, the build container remains started. Connecting to it, in `build` there is the project fetched from Github. 
- I tried also with default `abstruse_builder` and the simplest `.abstruse.yml` file I can imagine:

```
image: abstruse_builder

matrix:
  - env: NODE_VERSION=8

install:
  - echo "test"

script:
  - echo "test"
```

**Comment by jkuri** [comment]
do not use abstruse_builder as testing image, this is a base image only.

**Comment by jkuri** [comment]
can you try with our docker image or use node at least 9.0.0 please, seems
like an issue with dockerode library

**Comment by marcopiraccini** [comment]
...same with node 10. So tried with `bleenco/abstruse`, but: 
```
==> starting container abstruse_3_3 from image bleenco/abstruse ... done.
==> git clone -q https://github.com/marcopiraccini/ethplatform.git -b master .
OCI runtime exec failed: exec failed: container_linux.go:348: starting container process caused "exec: \"/usr/bin/abstruse-pty\": stat /usr/bin/abstruse-pty: no such file or directory": unknown

[error]: executed command returned exit code 255
```

**Comment by jkuri** [comment]
did you read the documentation? You need to build image using web UI.

**Comment by marcopiraccini** [comment]
Yes, I did.  I was just try to follow your suggestion: `can you try with our docker image`. 
So now I'm confused: what is "our docker image" if it's not:
- `abstruse`builder`
- a custom image build using the UI which extends `abstruse_builder`
- the `abstruse` image
?

However I think the problem is not in the image, it's something happening between Dockerode and Docker. Which Docker version are you using for testing?

**Comment by marcopiraccini** [comment]
Not sure why you marked as "question" since there's at least a configuration on which, for some unknown reason, this error (see below) is produced and the application crashes. 
The flow was already clear, but I agree with the fact that the cause must be "external" from the code. The last test has been done with:
```
root@xxx:~# docker -v
Docker version 18.06.0-ce-dev, build b4b92d0
root@xxx:~# node -v
v9.11.1
```
...and again same behaviour. 
I'm now investigating on the environment, which is a Digital Ocean Droplet. Maybe the resource limitation is affecting the nodejs <-> docker communication in some way. I will post here what I find. 

Here (again) the error: 
```
Error [ERR_STREAM_WRITE_AFTER_END]: write after end
    at writeAfterEnd (_stream_writable.js:241:12)
    at Writable.write (_stream_writable.js:291:5)
    at IncomingMessage.<anonymous> (/root/.nvm/versions/node/v9.11.1/lib/node_modules/abstruse/node_modules/docker-modem/lib/modem.js:301:16)
    at IncomingMessage.emit (events.js:180:13)
    at emitReadable_ (_stream_readable.js:525:10)
    at emitReadable (_stream_readable.js:519:7)
    at addChunk (_stream_readable.js:285:7)
    at readableAddChunk (_stream_readable.js:261:11)
    at IncomingMessage.Readable.push (_stream_readable.js:218:10)
    at HTTPParser.parserOnBody (_http_common.js:135:22)
    at Socket.socketOnData (_http_client.js:445:20)
    at Socket.emit (events.js:180:13)
    at addChunk (_stream_readable.js:274:12)
    at readableAddChunk (_stream_readable.js:261:11)
    at Socket.Readable.push (_stream_readable.js:218:10)
    at Pipe.onread (net.js:581:20)
Emitted 'error' event at:
    at writeAfterEnd (_stream_writable.js:243:10)
    at Writable.write (_stream_writable.js:291:5)
    [... lines matching original stack trace ...]
    at Pipe.onread (net.js:581:20)
```

**Comment by jkuri** [comment]
Hi Marco. Okay, please report what you will find out and thanks for the
reported error. I think the best I can do at this moment is to report bug
to dockerode, it is definitely error from there and not abstruse.

**Comment by jkuri** [comment]
@marcopiraccini any updates?

**Comment by marcopiraccini** [comment]
Not yet. I suspect it's related to https://github.com/apocas/dockerode/issues/422
..but no time to check it (yet).

**Comment by marcopiraccini** [comment]
OK, found something. 
Having the SAME EXACT version of Docker (now testing with `18.05.0-ce` helped a lot. 
(every like 10 builds it still happens...but now it's much much better...)

**Comment by irmana** [comment]
@marcopiraccini how about now, any updates? ;)

**Comment by marcopiraccini** [comment]
Hey :)
Not using right now, but I will in next days and then I'll check.

**Comment by marcopiraccini** [comment]
Nope, still the same. 
Tried `v1.6.4` using docker in:
- Ubuntu 16.04 (digital ocean), Ubuntu 18.04 (digital ocean)
- Ubuntu 16.04 (aws)
- Ubuntu 16.04 (my laptop, I was suspecting was a resources issue, so run with 16GB RAM).

Next step: not using Ubuntu, trying to understand if the issue is between the OS and docker

**Comment by jkuri** [comment]
Ubuntu should be fine.

**Comment by jkuri** [comment]
This issue was solved with https://github.com/bleenco/abstruse/commit/12c0956f9851d47ad99f555b5f9566b5164db578 but it is not included in `1.6.4` release. Please try building from master.

**Comment by marcopiraccini** [comment]
Tnx @jkuri , now the issue seems actually fixed!
Closing this issue.

**Comment by jkuri** [comment]
no problem, thanks for checking. this will be included in 2.0 release which is following.

**Comment by alexisrolland** [comment]
Hi guys, I'm facing a similar error message while using Dockerode in a Node.js. I'm wondering how you managed to solve it? https://stackoverflow.com/questions/59985046/error-err-stream-write-after-end-in-dockerode
